File: assignment_helper/rng.py
"""Counter-based PRNG. Invariant I14.

rand(block_seed, purpose, index) -> [0, 1)

Stateless by construction. NEVER a stateful stream: with a stream, inserting one
new jitter call shifts every downstream value and silently re-rolls the whole page.

This file has a bit-for-bit twin at web/src/render/rng.ts. The golden vectors in
tests/unit/test_rng_golden.py are shared by both and must never diverge.
"""

from __future__ import annotations

MASK64 = (1 << 64) - 1
_FNV_OFFSET = 0xCBF29CE484222325
_FNV_PRIME = 0x100000001B3


def fnv1a64(s: str) -> int:
    """FNV-1a over the UTF-8 bytes of s."""
    h = _FNV_OFFSET
    for b in s.encode("utf-8"):
        h = ((h ^ b) * _FNV_PRIME) & MASK64
    return h


def splitmix64(x: int) -> int:
    z = (x + 0x9E3779B97F4A7C15) & MASK64
    z = ((z ^ (z >> 30)) * 0xBF58476D1CE4E5B9) & MASK64
    z = ((z ^ (z >> 27)) * 0x94D049BB133111EB) & MASK64
    return (z ^ (z >> 31)) & MASK64


def rand_bits(block_seed: int, purpose: str, index: int) -> int:
    """The raw 64-bit draw. Exposed so the golden-vector test can compare exactly."""
    h = (block_seed & MASK64) ^ fnv1a64(purpose) ^ splitmix64(index & MASK64)
    return splitmix64(h & MASK64)


def rand(block_seed: int, purpose: str, index: int) -> float:
    """Uniform in [0, 1). Top 53 bits, so it is exactly representable as a float."""
    return (rand_bits(block_seed, purpose, index) >> 11) / float(1 << 53)


def rand_range(block_seed: int, purpose: str, index: int, lo: float, hi: float) -> float:
    return lo + (hi - lo) * rand(block_seed, purpose, index)


def rand_int(block_seed: int, purpose: str, index: int, n: int) -> int:
    """Uniform in [0, n). n must be positive."""
    if n <= 0:
        raise ValueError(f"rand_int needs a positive n, got {n}")
    return rand_bits(block_seed, purpose, index) % n

File: web/src/render/rng.ts
/**
 * Counter-based PRNG. Invariant I14.
 *
 *   rand(blockSeed, purpose, index) -> [0, 1)
 *
 * Stateless by construction. NEVER a stateful stream: with a stream, inserting one
 * new jitter call shifts every downstream value and silently re-rolls the whole page.
 *
 * This file has a bit-for-bit twin at assignment_helper/rng.py. The golden vectors in
 * tests/unit/rng.golden.json are shared by both and must never diverge.
 *
 * Leaf module: zero imports, by design (plan §2 module table).
 */

const MASK64 = (1n << 64n) - 1n;
const FNV_OFFSET = 0xcbf29ce484222325n;
const FNV_PRIME = 0x100000001b3n;

const encoder = new TextEncoder();

export function fnv1a64(s: string): bigint {
  let h = FNV_OFFSET;
  for (const b of encoder.encode(s)) {
    h = ((h ^ BigInt(b)) * FNV_PRIME) & MASK64;
  }
  return h;
}

export function splitmix64(x: bigint): bigint {
  let z = (x + 0x9e3779b97f4a7c15n) & MASK64;
  z = ((z ^ (z >> 30n)) * 0xbf58476d1ce4e5b9n) & MASK64;
  z = ((z ^ (z >> 27n)) * 0x94d049bb133111ebn) & MASK64;
  return (z ^ (z >> 31n)) & MASK64;
}

export function randBits(blockSeed: bigint, purpose: string, index: number): bigint {
  const h = ((blockSeed & MASK64) ^ fnv1a64(purpose) ^ splitmix64(BigInt(index) & MASK64)) & MASK64;
  return splitmix64(h);
}

/** Uniform in [0, 1). Top 53 bits, so it is exactly representable as a double. */
export function rand(blockSeed: bigint, purpose: string, index: number): number {
  return Number(randBits(blockSeed, purpose, index) >> 11n) / 9007199254740992; // 2^53
}

export function randRange(
  blockSeed: bigint,
  purpose: string,
  index: number,
  lo: number,
  hi: number,
): number {
  return lo + (hi - lo) * rand(blockSeed, purpose, index);
}

/** Uniform in [0, n). n must be a positive integer. */
export function randInt(blockSeed: bigint, purpose: string, index: number, n: number): number {
  if (!Number.isInteger(n) || n <= 0) {
    throw new Error(`randInt needs a positive integer n, got ${n}`);
  }
  return Number(randBits(blockSeed, purpose, index) % BigInt(n));
}

/**
 * Signed symmetric draw in [-amp, +amp]. The workhorse for jitter: it is centred,
 * so raising the neatness dial to its clean pole genuinely removes the effect
 * rather than biasing the page in one direction.
 */
export function randSigned(
  blockSeed: bigint,
  purpose: string,
  index: number,
  amp: number,
): number {
  return (rand(blockSeed, purpose, index) * 2 - 1) * amp;
}
